lib-ee/emqx_license/src/emqx_license_http_api.erl

```diff
@@ -0,0 +1,142 @@
+                                start_at => "2022-01-11",
+                                type => "trial"
+                            }
+                        }
+                    }
+                )
+            }
+        }
+    };
+schema("/license/upload") ->
+    #{
+        'operationId' => '/license/upload',
+        post => #{
+            tags => [<<"license">>],
+            summary => <<"Upload license">>,
+            description => ?DESC("desc_license_upload_api"),
+            'requestBody' => emqx_dashboard_swagger:schema_with_examples(
+                emqx_license_schema:license_type(),
+                #{
+                    license_key => #{
+                        summary => <<"License key string">>,
+                        value => #{
+                            <<"key">> => <<"xxx">>,
+                            <<"connection_low_watermark">> => "75%",
+                            <<"connection_high_watermark">> => "80%"
+                        }
+                    },
+                    license_file => #{
+                        summary => <<"Path to a license file">>,
+                        value => #{
+                            <<"file">> => <<"/path/to/license">>,
+                            <<"connection_low_watermark">> => "75%",
+                            <<"connection_high_watermark">> => "80%"
+                        }
+                    }
+                }
+            ),
+            responses => #{
+                200 => <<"ok">>,
+                400 => emqx_dashboard_swagger:error_codes([?BAD_REQUEST], <<"bad request">>),
+                404 => emqx_dashboard_swagger:error_codes([?NOT_FOUND], <<"file not found">>)
+            }
+        }
+    }.
+
+'/license'(get, _Params) ->
+    License = maps:from_list(emqx_license_checker:dump()),
+    {200, License}.
+
+'/license/upload'(post, #{body := #{<<"file">> := Filepath}}) ->
+    case emqx_license:update_file(Filepath) of
+        {error, enoent} ->
+            ?SLOG(error, #{
+                msg => "license_file_not_found",
+                path => Filepath
+            }),
+            {404, <<"file not found">>};
+        {error, Error} ->
+            ?SLOG(error, #{
+                msg => "bad_license_file",
+                reason => Error,
+                path => Filepath
+            }),
+            {400, <<"bad request">>};
+        {ok, _} ->
+            ?SLOG(info, #{
+                msg => "updated_license_file",
+                path => Filepath
+            }),
+            {200, <<"ok">>}
+    end;
+'/license/upload'(post, #{body := #{<<"key">> := Key}}) ->
+    case emqx_license:update_key(Key) of
+        {error, Error} ->
+            ?SLOG(error, #{
+                msg => "bad_license_key",
+                reason => Error
+            }),
+            {400, <<"bad request">>};
+        {ok, _} ->
+            ?SLOG(info, #{msg => "updated_license_key"}),
+            {200, <<"ok">>}
```

Comment:
```erlang
{200, #{<<"key">> => Key}}
```
Maybe response the created resource is better.

Reply:
I've changed it to return the license here: https://github.com/emqx/emqx/pull/8610/commits/50dc90596aa9f22def5d41b7376440320f6d69f7